beets/util/__init__.py

```diff
@@ -289,7 +290,7 @@ def fnmatch_all(names: Sequence[bytes], patterns: Sequence[bytes]) -> bool:
 
 def prune_dirs(
     path: str,
-    root: Optional[Bytes_or_String] = None,
+    root: Optional[AnyStr] = None,
```

Comment:
See my comments above: in summary, I think we'd rather use types that reflect the current implementation. In addition to the two reasons I've given above, I see another risk: if we were to restrict the accepted type to `bytes` here, in a fully-typed codebase we'd expect `mypy` check to fail anywhere where a `str` is being fed into this function.

Right now, much the codebase is not typed, so the check passes silently. Whoever ends up adding types to the part of the codebase that sends `str` into this function will get slightly bamboozled: they will need to read this code and understand what it does and they will have to either 

1. Adjust the implementation of that code
2. Relax the type here

Both of these actions will be out of scope of what they're doing. Remember, I had to do this in `playlist.py` when I ended up relaxing the type in `syspath`?

Reply:
I'd be happy with making these functions to handle `bytes` only if

1. We explicitly handle `bytes` only - remove the code that casts `str` to `bytes` from the beginning of the implementation
2. Rename the functions to make the _only-bytes-accepted-here_ requirement clear, since the equivalents in `os` (`normpath`, `samefile`, ...) handle `str` and `bytes` types. Therefore, to not confuse people used to the builtins, we'd want to name these functions accordingly I suppose.